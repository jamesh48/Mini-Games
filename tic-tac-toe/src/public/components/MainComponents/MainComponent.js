import React from 'react';
// Player Header
import { PlayerHeaderStoreProvider } from 'Components/PlayerHeaderComponents/PlayerHeaderStore/playerHeaderStore.js';
import PlayerHeaders from 'Components/PlayerHeaderComponents/PlayerHeaders.js';

// Board
import { BoardStoreProvider } from 'Components/BoardComponents/BoardStore/boardStore.js';
import Board from 'Components/BoardComponents/Board.js';

import PlayerNameEntry from 'Components/PlayerNameEntryComponents/PlayerNameEntry.js';

import './main-styles/reset.scss';
import './main-styles/main.scss';
import './main-styles/global.scss';

export default () => {
  return (
    <div id='container'>
      <PlayerHeaderStoreProvider>
        <PlayerHeaders />
      </PlayerHeaderStoreProvider>

      <BoardStoreProvider>
        <Board />
      </BoardStoreProvider>

      <PlayerNameEntry />
    </div>
  )
}