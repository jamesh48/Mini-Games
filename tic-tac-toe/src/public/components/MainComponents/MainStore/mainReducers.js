const reduceReducers = (...reducers) => {
  return (state, action) => reducers.reduce((acc, nextReducer) => nextReducer(acc, action), state)
};

const hasAlreadyWon = (state = null, action) => {
  switch (action.type) {
    case 'TOGGLE HAS ALREADY WON TRUE':
      return action.payload;
    default:
      return state;
  };
};


const xPlayerName = (state = null, action) => {
  switch (action.type) {
    case 'SUBMIT X PLAYER NAME':
      return action.payload;
    default:
      return state;
  };
};

const oPlayerName = (state = null, action) => {
  switch (action.type) {
    case 'SUBMIT O PLAYER NAME':
      return action.payload;
    default:
      return state;
  }
}

const combineReducers = (slices) => (state, action) => Object.keys(slices).reduce((acc, prop) => ({
  ...acc,
  [prop]: slices[prop](acc[prop], action),
}),
  state
);

export default combineReducers({ hasAlreadyWon, xPlayerName, oPlayerName });

