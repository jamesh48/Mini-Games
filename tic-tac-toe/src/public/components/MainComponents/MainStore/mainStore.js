// https://stackoverflow.com/questions/59200785/react-usereducer-how-to-combine-multiple-reducers
import React, { useReducer } from 'react';
import CombinedReducers from './mainReducers.js';
const MainStoreContext = React.createContext();

const initialState = {
  hasAlreadyWon: null,
  xPlayerName: null,
  oPlayerName: null
};

const MainStoreProvider = ({ children }) => {
  const [state, dispatch] = React.useReducer(CombinedReducers, initialState);
  const store = React.useMemo(() => [state, dispatch], [state]);
  return (
    <MainStoreContext.Provider value={store}>{children}</MainStoreContext.Provider>
  );
};

const useMainContext = () => {
  const context = React.useContext(MainStoreContext);
  if (context === undefined) {
    throw new Error('useMainStoreContext must be used within a MainStoreProvider')
  }
  return context;
};


export { useMainContext, MainStoreProvider };