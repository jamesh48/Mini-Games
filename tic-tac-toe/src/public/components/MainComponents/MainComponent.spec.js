import React from 'react';
import { MainStoreProvider } from 'MainStore';
import MainComponent from 'Components/MainComponents/MainComponent';

import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

describe('Main Component', () => {
  it('Renders the Main Component', () => {
    const wrapper = mount(
      <MainStoreProvider>
        <MainComponent />
      </MainStoreProvider>
    );
    expect(wrapper.find('#container').length).toEqual(1);
  });

  test('snapshot renders', () => {
    const component = renderer.create(
      <MainStoreProvider>
        <MainComponent />
      </MainStoreProvider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});