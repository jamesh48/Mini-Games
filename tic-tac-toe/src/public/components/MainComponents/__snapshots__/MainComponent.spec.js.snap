// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Component snapshot renders 1`] = `
<div
  id="container"
>
  <div
    id="player-headers"
  >
    <span
      className="player-name-headers"
      id="p1-name-header"
    >
      X Player-&gt; 
      <h1
        className="player-name-tally"
        id="p-x-tally"
      >
        0
      </h1>
    </span>
    <span
      className="player-name-headers"
      id="p2-name-header"
    >
      O Player-&gt; 
      <p
        className="player-name-tally"
        id="p-o-tally"
      >
        0
      </p>
    </span>
  </div>
  <div
    id="board"
  >
    <div
      className="board-row"
    >
      <span
        className="boxes"
        id="box-0"
        onClick={[Function]}
      />
      <span
        className="boxes"
        id="box-1"
        onClick={[Function]}
      />
      <span
        className="boxes"
        id="box-2"
        onClick={[Function]}
      />
    </div>
    <div
      className="board-row"
    >
      <span
        className="boxes"
        id="box-3"
        onClick={[Function]}
      />
      <span
        className="boxes"
        id="box-4"
        onClick={[Function]}
      />
      <span
        className="boxes"
        id="box-5"
        onClick={[Function]}
      />
    </div>
    <div
      className="board-row"
    >
      <span
        className="boxes"
        id="box-6"
        onClick={[Function]}
      />
      <span
        className="boxes"
        id="box-7"
        onClick={[Function]}
      />
      <span
        className="boxes"
        id="box-8"
        onClick={[Function]}
      />
    </div>
  </div>
  <div
    id="player-name-entry"
  >
    <legend
      id="enter-your-players"
    >
      Enter Your Players!
    </legend>
    <form
      className="enter-players"
      onSubmit={[Function]}
    >
      <input
        autoComplete="off"
        className="player-name-input"
        id="p1"
        onChange={[Function]}
        placeholder="X Player"
        type="text"
        value=""
      />
      <input
        autoComplete="off"
        className="player-name-input"
        id="p2"
        onChange={[Function]}
        placeholder="O Player"
        type="text"
        value=""
      />
      <input
        autoComplete="off"
        className="player-name-input"
        id="board-size"
        placeholder="Board Width"
        type="text"
      />
      <button
        hidden={true}
        type="submit"
        value="Submit"
      >
        Submit Form
      </button>
    </form>
    <form>
      <input
        id="new-game-button"
        type="button"
        value="New Game!"
      />
    </form>
  </div>
</div>
`;
